File: app_python/mod_dimension.py
# =============================================================================
# mod_dimension.py — Module générique : Couverture / Infrastructures / Enseignants
# =============================================================================
from datetime import date

import pandas as pd
from shiny import module, ui, render, reactive, req
from shinywidgets import output_widget, render_widget
from faicons import icon_svg

import commun as C

CONFIG = {
    "couverture": dict(
        titre="Couverture scolaire", icone="school",
        variables_carte={
            "nb_etab": "Nombre d'établissements",
            "densite_etab_100km2": "Établissements pour 100 km²",
            "completude_offre": "Complétude du cycle éducatif (%)",
            "jardins_100_primaires": "Jardins d'enfants pour 100 primaires",
            "nb_primaire": "Écoles primaires", "nb_college": "Collèges", "nb_lycee": "Lycées"},
        variables_comparaison=["nb_etab", "densite_etab_100km2", "completude_offre",
                               "jardins_100_primaires", "nb_prescolaire", "nb_primaire",
                               "nb_college", "nb_lycee"]),
    "infrastructures": dict(
        titre="Infrastructures essentielles", icone="building",
        variables_carte={
            "toilettes_100_etab": "Toilettes pour 100 établissements",
            "pct_toilettes_ameliorees": "Toilettes améliorées (%)",
            "pct_batiments_recents": "Bâtiments récents — depuis 2010 (%)",
            "batiments_par_etab": "Bâtiments par établissement",
            "pct_terrain_sport": "Terrains de sport (%)",
            "biblio_100_etab": "Bibliothèques pour 100 établissements"},
        variables_comparaison=["toilettes_100_etab", "pct_toilettes_ameliorees",
                               "pct_batiments_recents", "batiments_par_etab",
                               "pct_terrain_sport", "biblio_100_etab"]),
    "enseignants": dict(
        titre="Capacité enseignante", icone="chalkboard-user",
        variables_carte={
            "enseignants_prescolaire": "Enseignants du préscolaire",
            "ens_100_jardins": "Enseignants pour 100 jardins d'enfants",
            "enseignants_100_etab": "Enseignants pour 100 établissements"},
        variables_comparaison=["enseignants_prescolaire", "ens_100_jardins",
                               "enseignants_100_etab", "nb_prescolaire"]),
}

_choix_niveaux = {"region": "Région", "prefecture": "Préfecture", "commune": "Commune"}


@module.ui
def dimension_ui(dim: str):
    cfg = CONFIG[dim]
    return ui.div(
        ui.output_ui("kpis"),
        ui.navset_card_pill(
            ui.nav_panel(
                "Indicateurs",
                ui.layout_columns(
                    ui.input_radio_buttons("niveau_ind", "Échelon territorial",
                                           _choix_niveaux, selected="prefecture", inline=True),
                    col_widths=[6]),
                ui.layout_columns(
                    output_widget("graphique1"), output_widget("graphique2"),
                    col_widths=[6, 6]),
                ui.output_data_frame("table_indicateurs"),
            ),
            ui.nav_panel(
                "Cartographie",
                ui.layout_columns(
                    ui.div(
                        ui.input_select("var_carte", "Indicateur cartographié",
                                        {k: v for k, v in cfg["variables_carte"].items()}),
                        ui.input_radio_buttons("niveau_carte", "Maille", _choix_niveaux,
                                               selected="commune", inline=True),
                        ui.input_selectize("filtre_region", "Filtrer par région",
                                           sorted(C.IND["commune"]["region"].unique().tolist()),
                                           multiple=True),
                        ui.input_switch("montrer_points", "Afficher les établissements", False),
                        ui.tags.hr(),
                        ui.tags.b("Télécharger la carte affichée"),
                        ui.tags.p("Avec filtres, légende et titre.",
                                  style="font-size:12px;color:#777;"),
                        ui.download_button("dl_png", "PNG haute résolution",
                                           class_="btn-outline-success btn-sm w-100 mb-1"),
                        ui.download_button("dl_pdf", "PDF", class_="btn-outline-success btn-sm w-100 mb-1"),
                        ui.download_button("dl_jpeg", "JPEG", class_="btn-outline-success btn-sm w-100"),
                    ),
                    ui.output_ui("carte"),
                    col_widths=[3, 9]),
            ),
            ui.nav_panel(
                "Comparaison",
                ui.layout_columns(
                    ui.div(
                        ui.input_radio_buttons("niveau_comp", "Échelon", _choix_niveaux,
                                               selected="prefecture", inline=True),
                        ui.output_ui("choix_comparaison")),
                    output_widget("radar_comp"),
                    col_widths=[4, 8]),
                output_widget("barres_comp"),
                ui.output_data_frame("table_comp"),
            ),
            ui.nav_panel(
                "Export",
                ui.layout_columns(
                    ui.div(
                        ui.input_radio_buttons("niveau_export", "Échelon à exporter",
                                               _choix_niveaux, selected="commune", inline=True),
                        ui.download_button("dl_excel", "Excel (.xlsx)", class_="btn-success w-100 mb-1"),
                        ui.download_button("dl_csv", "CSV", class_="btn-outline-success w-100")),
                    ui.output_data_frame("apercu_export"),
                    col_widths=[4, 8]),
            ),
            title=ui.span(icon_svg(cfg["icone"]), " " + cfg["titre"]),
            id="onglets",
        ),
    )


@module.server
def dimension_server(input, output, session, dim: str, onglet_actif=None):
    cfg = CONFIG[dim]
    vars_dim = [v for v in dict.fromkeys(cfg["variables_comparaison"]
                                          + list(cfg["variables_carte"]))
                if v in C.IND["commune"].columns]

    # ---------- KPIs ----------
    @render.ui
    def kpis():
        i = C.IND["commune"]
        if dim == "couverture":
            valeurs = [(C.fmt(C.META["nb_etablissements"]), "Établissements scolaires", "school"),
                       (C.fmt(i["nb_primaire"].sum()), "Écoles primaires", "children"),
                       (C.fmt1(i["completude_offre"].mean()) + " %", "Complétude moyenne du cycle", "layer-group"),
                       (C.fmt(C.META["nb_communes"]), "Communes couvertes", "map")]
        elif dim == "infrastructures":
            valeurs = [(C.fmt(C.META["nb_toilettes"]), "Points de toilettes recensés", "restroom"),
                       (C.fmt(C.META["nb_batiments"]), "Bâtiments scolaires", "building"),
                       (C.fmt1(i["pct_toilettes_ameliorees"].mean()) + " %", "Toilettes améliorées (moyenne)", "droplet"),
                       (C.fmt1(i["pct_terrain_sport"].mean()) + " %", "Établissements avec terrain de sport", "futbol")]
        else:
            p = C.IND["prefecture"]
            valeurs = [(C.fmt(C.META["nb_enseignants_prescolaire"]), "Enseignants du préscolaire", "chalkboard-user"),
                       (C.fmt1(p["ens_100_jardins"].median()), "Enseignants / 100 jardins (médiane)", "users"),
                       (C.fmt(i["nb_prescolaire"].sum()), "Jardins d'enfants", "shapes"),
                       (C.fmt((p["ens_100_jardins"] < p["ens_100_jardins"].quantile(.25)).sum()),
                        "Préfectures sous-dotées (quartile inf.)", "triangle-exclamation")]
        themes = ["success", "primary", "warning", "danger"]
        return ui.layout_columns(
            *[ui.value_box(t, v, showcase=icon_svg(ic), theme=th)
              for (v, t, ic), th in zip(valeurs, themes)],
            col_widths=[3, 3, 3, 3])

    # ---------- Indicateurs ----------
    @render_widget
    def graphique1():
        if dim == "couverture":
            rep = C.POINTS["categorie"].value_counts()
            import plotly.graph_objects as go
            fig = go.Figure(go.Pie(labels=rep.index, values=rep.values, hole=0))
            return C.theme_plotly(fig, "Répartition nationale par type d'établissement")
        if dim == "infrastructures":
            d = C.UIS_INFRA[C.UIS_INFRA["indicateur"].str.contains("lectrifi")]
            return C.series_temporelles(d, "indicateur",
                                        "Électrification des établissements (national, UNESCO-ISU)", "%")
        d = C.UIS_RATIO
        return C.series_temporelles(d, "indicateur",
                                    "Ratio élèves / enseignant (national, UNESCO-ISU)")

    @render_widget
    def graphique2():
        niv = input.niveau_ind()
        var = {"couverture": "nb_etab", "infrastructures": "toilettes_100_etab",
               "enseignants": "ens_100_jardins"}[dim]
        d = C.IND[niv].nlargest(20, var)
        titre = {"couverture": "Répartition territoriale des établissements",
                 "infrastructures": "Toilettes pour 100 établissements",
                 "enseignants": "Encadrement préscolaire par territoire"}[dim]
        return C.barres(d, niv, var, titre + (" (top 20)" if len(C.IND[niv]) > 20 else ""))

    @render.data_frame
    def table_indicateurs():
        niv = input.niveau_ind()
        d = C.IND[niv][C.cles(niv) + vars_dim].copy()
        d.columns = [C.NIVEAUX_LABELS.get(c, C.LIB.get(c, c)) for c in d.columns]
        return render.DataGrid(d, width="100%", height="380px", filters=True)

    # ---------- Cartographie ----------
    def couche_filtree():
        g = C.SF[input.niveau_carte()]
        regs = list(input.filtre_region())
        if regs:
            g = g[g["region"].isin(regs)]
        return g

    @render.ui
    def carte():
        g = couche_filtree()
        pts = None
        if input.montrer_points():
            pts = C.POINTS
            regs = list(input.filtre_region())
            if regs:
                pts = pts[pts["region"].isin(regs)]
        return ui.HTML(C.carte_leaflet(g, input.var_carte(),
                                       cfg["variables_carte"][input.var_carte()],
                                       input.niveau_carte(), points=pts))

    def _export_carte(fm):
        g = couche_filtree()
        regs = list(input.filtre_region())
        sous = ("Régions : " + ", ".join(regs)) if regs else "Ensemble du territoire national"
        return C.exporter_carte(
            g, input.var_carte(),
            f"{cfg['titre']} — {cfg['variables_carte'][input.var_carte()]}",
            f"{sous} · Maille : {C.NIVEAUX_LABELS[input.niveau_carte()]}", fm)

    @render.download(filename=lambda: f"carte_{dim}_{date.today():%Y%m%d}.png")
    def dl_png():
        yield _export_carte("png")

    @render.download(filename=lambda: f"carte_{dim}_{date.today():%Y%m%d}.pdf")
    def dl_pdf():
        yield _export_carte("pdf")

    @render.download(filename=lambda: f"carte_{dim}_{date.today():%Y%m%d}.jpeg")
    def dl_jpeg():
        yield _export_carte("jpeg")

    # ---------- Comparaison ----------
    @render.ui
    def choix_comparaison():
        niv = input.niveau_comp()
        return ui.input_selectize("territoires_comp", "Territoires à comparer",
                                  sorted(C.IND[niv][niv].tolist()), multiple=True)

    @reactive.calc
    def d_comp():
        niv = input.niveau_comp()
        sel = list(input.territoires_comp() or [])
        return C.IND[niv][C.IND[niv][niv].isin(sel)]

    @render_widget
    def radar_comp():
        niv = input.niveau_comp()
        d = d_comp()
        ens = C.IND[niv]
        axes = [C.LIB.get(v, v) for v in cfg["variables_comparaison"]]
        traces = []
        for _, ligne in d.iterrows():
            vals = [float((ens[v] <= ligne[v]).mean() * 100)
                    for v in cfg["variables_comparaison"]]
            traces.append((ligne[niv], vals))
        return C.radar(traces, axes, "Positionnement relatif (percentile national)")

    @render_widget
    def barres_comp():
        import plotly.graph_objects as go
        niv = input.niveau_comp()
        d = d_comp()
        fig = go.Figure()
        for i, (_, ligne) in enumerate(d.iterrows()):
            fig.add_trace(go.Bar(
                name=ligne[niv],
                x=[C.LIB.get(v, v) for v in cfg["variables_comparaison"]],
                y=[ligne[v] for v in cfg["variables_comparaison"]],
                marker_color=C.COULEURS_SERIES[i % 6]))
        fig.update_layout(barmode="group")
        return C.theme_plotly(fig, "Comparaison des indicateurs (valeurs brutes)")

    @render.data_frame
    def table_comp():
        niv = input.niveau_comp()
        d = d_comp()[C.cles(niv) + cfg["variables_comparaison"]].copy()
        d.columns = [C.NIVEAUX_LABELS.get(c, C.LIB.get(c, c)) for c in d.columns]
        return render.DataGrid(d, width="100%")

    # ---------- Export ----------
    def d_export():
        niv = input.niveau_export()
        d = C.IND[niv][C.cles(niv) + vars_dim].copy()
        d.columns = [C.NIVEAUX_LABELS.get(c, C.LIB.get(c, c)) for c in d.columns]
        return d

    @render.data_frame
    def apercu_export():
        return render.DataGrid(d_export(), width="100%", height="420px")

    @render.download(filename=lambda: f"{dim}_indicateurs_{date.today():%Y%m%d}.xlsx")
    def dl_excel():
        yield C.exporter_excel({cfg["titre"][:31]: d_export()})

    @render.download(filename=lambda: f"{dim}_indicateurs_{date.today():%Y%m%d}.csv")
    def dl_csv():
        yield C.exporter_csv(d_export())

File: app/R/mod_dimension.R
# =============================================================================
# mod_dimension.R — Module générique des pages Couverture scolaire,
# Infrastructures essentielles et Capacité enseignante.
# Chaque page : Indicateurs | Cartographie | Comparaison | Exports
# =============================================================================

CONFIG_DIM <- list(
  couverture = list(
    titre = "Couverture scolaire", icone = "school",
    variables_carte = c("Nombre d'établissements" = "nb_etab",
                        "Établissements pour 100 km²" = "densite_etab_100km2",
                        "Complétude du cycle éducatif (%)" = "completude_offre",
                        "Jardins d'enfants pour 100 primaires" = "jardins_100_primaires",
                        "Écoles primaires" = "nb_primaire",
                        "Collèges" = "nb_college",
                        "Lycées" = "nb_lycee",
                        "Crèches recensées" = "nb_creches"),
    variables_comparaison = c("nb_etab", "densite_etab_100km2", "completude_offre",
                              "jardins_100_primaires", "nb_prescolaire", "nb_primaire",
                              "nb_college", "nb_lycee"),
    couche_points = "points_etabs"
  ),
  infrastructures = list(
    titre = "Infrastructures essentielles", icone = "building",
    variables_carte = c("Toilettes pour 100 établissements" = "toilettes_100_etab",
                        "Toilettes améliorées (%)" = "pct_toilettes_ameliorees",
                        "Bâtiments récents - depuis 2010 (%)" = "pct_batiments_recents",
                        "Bâtiments par établissement" = "batiments_par_etab",
                        "Terrains de sport (%)" = "pct_terrain_sport",
                        "Bibliothèques pour 100 établissements" = "biblio_100_etab"),
    variables_comparaison = c("toilettes_100_etab", "pct_toilettes_ameliorees",
                              "pct_batiments_recents", "batiments_par_etab",
                              "pct_terrain_sport", "biblio_100_etab"),
    couche_points = "points_toilettes"
  ),
  enseignants = list(
    titre = "Capacité enseignante", icone = "chalkboard-user",
    variables_carte = c("Enseignants du préscolaire" = "enseignants_prescolaire",
                        "Enseignants pour 100 jardins d'enfants" = "ens_100_jardins",
                        "Enseignants pour 100 établissements" = "enseignants_100_etab"),
    variables_comparaison = c("enseignants_prescolaire", "ens_100_jardins",
                              "enseignants_100_etab", "nb_prescolaire"),
    couche_points = "points_etabs"
  )
)

mod_dimension_ui <- function(id) {
  ns <- NS(id)
  cfg <- CONFIG_DIM[[id]]

  tagList(
    fluidRow(
      column(3, bs4ValueBoxOutput(ns("kpi1"), width = NULL)),
      column(3, bs4ValueBoxOutput(ns("kpi2"), width = NULL)),
      column(3, bs4ValueBoxOutput(ns("kpi3"), width = NULL)),
      column(3, bs4ValueBoxOutput(ns("kpi4"), width = NULL))
    ),
    bs4TabCard(
      id = ns("onglets"), width = 12, collapsible = FALSE, maximizable = TRUE,
      status = "success", solidHeader = FALSE, type = "tabs", side = "right",
      title = tags$span(icon(cfg$icone), " ", cfg$titre),

      tabPanel(
        title = tagList(icon("chart-column"), "Indicateurs"),
        fluidRow(
          column(4,
                 radioGroupButtons(ns("niveau_ind"), "Échelon territorial",
                                   choices = c("Région" = "region", "Préfecture" = "prefecture",
                                               "Commune" = "commune"),
                                   selected = "prefecture", status = "success", justified = TRUE))
        ),
        fluidRow(
          column(6, highchartOutput(ns("graphique1"), height = 380)),
          column(6, highchartOutput(ns("graphique2"), height = 380))
        ),
        tags$hr(),
        DTOutput(ns("table_indicateurs"))
      ),

      tabPanel(
        title = tagList(icon("map-location-dot"), "Cartographie"),
        fluidRow(
          column(3,
                 selectInput(ns("var_carte"), "Indicateur cartographié",
                             choices = cfg$variables_carte, width = "100%"),
                 radioGroupButtons(ns("niveau_carte"), "Maille",
                                   choices = c("Région" = "region", "Préfecture" = "prefecture",
                                               "Commune" = "commune"),
                                   selected = "commune", status = "success", justified = TRUE),
                 pickerInput(ns("filtre_region"), "Filtrer par région",
                             choices = sort(unique(INDIC$commune$region)), multiple = TRUE,
                             options = pickerOptions(actionsBox = TRUE, liveSearch = TRUE,
                                                     noneSelectedText = "Tout le Togo")),
                 selectizeInput(ns("recherche"), "Rechercher un territoire",
                                choices = NULL, options = list(placeholder = "Nom du territoire...")),
                 materialSwitch(ns("montrer_points"), "Afficher les établissements",
                                value = FALSE, status = "success"),
                 tags$hr(),
                 tags$b("Télécharger la carte affichée"),
                 tags$p(class = "text-muted", style = "font-size:12px;",
                        "Conserve le zoom, les filtres, la légende et le titre."),
                 div(class = "btn-group-vertical", style = "width:100%;",
                     downloadButton(ns("dl_png"), "PNG haute résolution", class = "btn-outline-success btn-sm"),
                     downloadButton(ns("dl_pdf"), "PDF", class = "btn-outline-success btn-sm"),
                     downloadButton(ns("dl_jpeg"), "JPEG", class = "btn-outline-success btn-sm"))
          ),
          column(9, leafletOutput(ns("carte"), height = 620))
        )
      ),

      tabPanel(
        title = tagList(icon("scale-balanced"), "Comparaison"),
        fluidRow(
          column(4,
                 radioGroupButtons(ns("niveau_comp"), "Échelon",
                                   choices = c("Région" = "region", "Préfecture" = "prefecture",
                                               "Commune" = "commune"),
                                   selected = "prefecture", status = "success", justified = TRUE),
                 uiOutput(ns("choix_comparaison"))),
          column(8, plotlyOutput(ns("radar"), height = 420))
        ),
        tags$hr(),
        highchartOutput(ns("barres_comparaison"), height = 380),
        DTOutput(ns("table_comparaison"))
      ),

      tabPanel(
        title = tagList(icon("download"), "Export"),
        fluidRow(
          column(4,
                 radioGroupButtons(ns("niveau_export"), "Échelon à exporter",
                                   choices = c("Région" = "region", "Préfecture" = "prefecture",
                                               "Commune" = "commune"),
                                   selected = "commune", status = "success", justified = TRUE),
                 tags$p(class = "text-muted", "Le fichier contient tous les indicateurs de la dimension."),
                 div(class = "btn-group-vertical", style = "width:100%;",
                     downloadButton(ns("dl_excel"), "Excel (.xlsx)", class = "btn-success"),
                     downloadButton(ns("dl_csv"), "CSV", class = "btn-outline-success"),
                     downloadButton(ns("dl_pdf_tab"), "PDF", class = "btn-outline-success"))),
          column(8, DTOutput(ns("apercu_export")))
        )
      )
    )
  )
}

mod_dimension_server <- function(id) {
  moduleServer(id, function(input, output, session) {
    ns <- session$ns
    cfg <- CONFIG_DIM[[id]]
    vars_dim <- intersect(unique(c(cfg$variables_comparaison, unname(cfg$variables_carte))),
                          names(INDIC$commune))

    # ---------------- KPIs ----------------
    output$kpi1 <- renderbs4ValueBox({
      switch(id,
        couverture = boite_valeur(fmt(META$nb_etablissements), "Établissements scolaires", "school", "success"),
        infrastructures = boite_valeur(fmt(META$nb_toilettes), "Points de toilettes recensés", "restroom", "success"),
        enseignants = boite_valeur(fmt(META$nb_enseignants_prescolaire), "Enseignants du préscolaire", "chalkboard-user", "success"))
    })
    output$kpi2 <- renderbs4ValueBox({
      switch(id,
        couverture = boite_valeur(fmt(sum(INDIC$commune$nb_primaire)), "Écoles primaires", "children", "primary"),
        infrastructures = boite_valeur(fmt(META$nb_batiments), "Bâtiments scolaires", "building", "primary"),
        enseignants = boite_valeur(fmt1(median(INDIC$prefecture$ens_100_jardins, na.rm = TRUE)),
                                   "Enseignants / 100 jardins (médiane)", "users", "primary"))
    })
    output$kpi3 <- renderbs4ValueBox({
      switch(id,
        couverture = boite_valeur(paste0(fmt1(mean(INDIC$commune$completude_offre)), " %"),
                                  "Complétude moyenne du cycle", "layer-group", "warning"),
        infrastructures = boite_valeur(paste0(fmt1(mean(INDIC$commune$pct_toilettes_ameliorees, na.rm = TRUE)), " %"),
                                       "Toilettes améliorées (moyenne)", "droplet", "warning"),
        enseignants = boite_valeur(fmt(sum(INDIC$commune$nb_prescolaire)), "Jardins d'enfants", "shapes", "warning"))
    })
    output$kpi4 <- renderbs4ValueBox({
      switch(id,
        couverture = boite_valeur(fmt(META$nb_communes), "Communes couvertes", "map", "danger"),
        infrastructures = boite_valeur(paste0(fmt1(mean(INDIC$commune$pct_terrain_sport, na.rm = TRUE)), " %"),
                                       "Établissements avec terrain de sport", "futbol", "danger"),
        enseignants = boite_valeur(
          fmt(sum(INDIC$prefecture$ens_100_jardins < quantile(INDIC$prefecture$ens_100_jardins, .25, na.rm = TRUE), na.rm = TRUE)),
          "Préfectures sous-dotées (quartile inférieur)", "triangle-exclamation", "danger"))
    })

    # ---------------- Indicateurs : graphiques ----------------
    donnees_niveau <- reactive(INDIC[[input$niveau_ind]])

    output$graphique1 <- renderHighchart({
      d <- donnees_niveau(); niv <- input$niveau_ind
      if (id == "couverture") {
        rep_cat <- BASE$etablissements |> count(categorie)
        hchart(rep_cat, "pie", hcaes(name = categorie, y = n), name = "Établissements") |>
          hc_title(text = "Répartition nationale par type d'établissement") |>
          hc_add_theme(hc_theme_togo)
      } else if (id == "infrastructures") {
        d2 <- RESNAT$uis_infrastructures |>
          filter(grepl("lectrifi", indicateur)) |> arrange(annee)
        hc_serie_temporelle(d2, "indicateur", "Électrification des établissements (national, UNESCO-ISU)", "%")
      } else {
        d2 <- BASE$enseignants_sexe |> group_by(sexe) |> summarise(enseignants = sum(enseignants))
        hchart(d2, "pie", hcaes(name = sexe, y = enseignants), name = "Enseignants") |>
          hc_title(text = "Enseignants du préscolaire par sexe (2021-2022)") |>
          hc_add_theme(hc_theme_togo)
      }
    })

    output$graphique2 <- renderHighchart({
      d <- donnees_niveau(); niv <- input$niveau_ind
      var <- switch(id, couverture = "nb_etab", infrastructures = "toilettes_100_etab",
                    enseignants = "ens_100_jardins")
      titre <- switch(id,
        couverture = "Répartition territoriale des établissements",
        infrastructures = "Toilettes pour 100 établissements par territoire",
        enseignants = "Encadrement préscolaire par territoire")
      d <- d |> arrange(desc(.data[[var]])) |> head(20) |>
        mutate(territoire = .data[[niv]], valeur = .data[[var]])
      hchart(d, "bar", hcaes(x = territoire, y = valeur), name = LIB[[var]],
             color = TG_VERT) |>
        hc_title(text = paste0(titre, if (nrow(donnees_niveau()) > 20) " (top 20)" else "")) |>
        hc_xAxis(title = list(text = NULL)) |>
        hc_yAxis(title = list(text = LIB[[var]])) |>
        hc_add_theme(hc_theme_togo)
    })

    output$table_indicateurs <- renderDT({
      niv <- input$niveau_ind
      cles <- intersect(c("region", "prefecture", "commune"), names(donnees_niveau()))
      d <- donnees_niveau()[, c(cles, vars_dim)]
      colnames(d) <- c(NIVEAUX_LABELS[cles], unname(LIB[vars_dim]))
      datatable(d, rownames = FALSE, extensions = "Buttons",
                options = list(pageLength = 10, scrollX = TRUE, dom = "ftip"),
                class = "table-striped compact") |>
        formatStyle(columns = seq_along(d), fontSize = "13px")
    })

    # ---------------- Cartographie ----------------
    couche_carte <- reactive({
      couche <- SF[[input$niveau_carte]]
      if (length(input$filtre_region) > 0) couche <- couche |> filter(region %in% input$filtre_region)
      couche
    })

    observe({
      noms <- sort(nom_terr(sf::st_drop_geometry(couche_carte()), input$niveau_carte))
      updateSelectizeInput(session, "recherche", choices = c("", noms), server = TRUE)
    })

    output$carte <- renderLeaflet({
      couche <- couche_carte()
      lib_var <- names(cfg$variables_carte)[cfg$variables_carte == input$var_carte]
      m <- carte_base() |>
        ajouter_choroplethe(couche, input$var_carte, lib_var, niveau = input$niveau_carte)
      if (isTRUE(input$montrer_points)) {
        pts <- CARTES[[cfg$couche_points]]
        if (length(input$filtre_region) > 0) pts <- pts |> filter(region %in% input$filtre_region)
        if (nrow(pts) > 0)
          m <- m |> addCircleMarkers(data = pts, lng = ~lon, lat = ~lat, radius = 2.5,
                                     stroke = FALSE, fillOpacity = 0.55, fillColor = TG_ROUGE,
                                     clusterOptions = markerClusterOptions(maxClusterRadius = 45),
                                     label = if ("nom" %in% names(pts)) ~nom else NULL,
                                     group = "Établissements")
      }
      m
    })

    observeEvent(input$recherche, {
      req(nzchar(input$recherche))
      couche <- couche_carte()
      sel <- couche[nom_terr(sf::st_drop_geometry(couche), input$niveau_carte) == input$recherche, ]
      req(nrow(sel) == 1)
      bb <- as.numeric(sf::st_bbox(sel))
      leafletProxy("carte") |> fitBounds(bb[1], bb[2], bb[3], bb[4])
    })

    dessin_courant <- function() {
      couche <- couche_carte()
      lib_var <- names(cfg$variables_carte)[cfg$variables_carte == input$var_carte]
      filtres <- if (length(input$filtre_region) > 0)
        paste("Régions :", paste(input$filtre_region, collapse = ", ")) else "Ensemble du territoire national"
      pts <- NULL
      if (isTRUE(input$montrer_points)) {
        pts <- CARTES[[cfg$couche_points]]
        if (length(input$filtre_region) > 0) pts <- pts |> filter(region %in% input$filtre_region)
      }
      function() dessiner_carte_statique(
        couche, input$var_carte,
        titre = sprintf("%s — %s", cfg$titre, lib_var),
        sous_titre = sprintf("%s · Maille : %s", filtres, NIVEAUX_LABELS[[input$niveau_carte]]),
        bounds = input$carte_bounds, points = pts)
    }
    output$dl_png  <- telecharger_carte("png",  function() dessin_courant()(), paste0("carte_", id))
    output$dl_pdf  <- telecharger_carte("pdf",  function() dessin_courant()(), paste0("carte_", id))
    output$dl_jpeg <- telecharger_carte("jpeg", function() dessin_courant()(), paste0("carte_", id))

    # ---------------- Comparaison ----------------
    output$choix_comparaison <- renderUI({
      selecteur_territoires(ns("territoires_comp"), niveau = input$niveau_comp)
    })

    donnees_comp <- reactive({
      req(input$territoires_comp)
      niv <- input$niveau_comp
      INDIC[[niv]] |> filter(.data[[niv]] %in% input$territoires_comp)
    })

    output$radar <- renderPlotly({
      d <- donnees_comp(); niv <- input$niveau_comp
      req(nrow(d) > 0)
      ens <- INDIC[[niv]]
      fig <- plot_ly(type = "scatterpolar", mode = "lines+markers", fill = "toself")
      for (i in seq_len(nrow(d))) {
        vals <- vapply(cfg$variables_comparaison, function(v) {
          x <- ens[[v]]; r <- rank(x, ties.method = "average") / length(x) * 100
          r[which(nom_terr(ens, niv) == nom_terr(d, niv)[i])]
        }, numeric(1))
        fig <- fig |> add_trace(r = c(vals, vals[1]),
                                theta = c(unname(LIB[cfg$variables_comparaison]),
                                          LIB[[cfg$variables_comparaison[1]]]),
                                name = nom_terr(d, niv)[i], opacity = 0.6)
      }
      fig |> layout(polar = list(radialaxis = list(range = c(0, 100), ticksuffix = "")),
                    legend = list(orientation = "h"),
                    margin = list(t = 30),
                    title = list(text = "Positionnement relatif (percentile national)", font = list(size = 14)))
    })

    output$barres_comparaison <- renderHighchart({
      d <- donnees_comp(); niv <- input$niveau_comp
      req(nrow(d) > 0)
      dl <- d |> select(all_of(c(niv, cfg$variables_comparaison))) |>
        pivot_longer(-all_of(niv), names_to = "variable", values_to = "valeur") |>
        mutate(variable = unname(LIB[variable]), territoire = .data[[niv]])
      hchart(dl, "column", hcaes(x = variable, y = valeur, group = territoire)) |>
        hc_title(text = "Comparaison des indicateurs (valeurs brutes)") |>
        hc_xAxis(title = list(text = NULL)) |> hc_yAxis(title = list(text = NULL)) |>
        hc_add_theme(hc_theme_togo)
    })

    output$table_comparaison <- renderDT({
      d <- donnees_comp(); niv <- input$niveau_comp
      cles <- intersect(c("region", "prefecture", "commune"), names(d))
      dd <- d[, c(cles, cfg$variables_comparaison)]
      colnames(dd) <- c(NIVEAUX_LABELS[cles], unname(LIB[cfg$variables_comparaison]))
      datatable(dd, rownames = FALSE, options = list(dom = "t", scrollX = TRUE),
                class = "table-striped compact")
    })

    # ---------------- Exports ----------------
    donnees_export <- reactive({
      niv <- input$niveau_export
      cles <- intersect(c("region", "prefecture", "commune"), names(INDIC[[niv]]))
      d <- INDIC[[niv]][, c(cles, vars_dim)]
      colnames(d) <- c(NIVEAUX_LABELS[cles], unname(LIB[vars_dim]))
      d
    })
    output$apercu_export <- renderDT(
      datatable(donnees_export(), rownames = FALSE,
                options = list(pageLength = 8, scrollX = TRUE, dom = "tip"),
                class = "table-striped compact"))
    output$dl_excel <- telecharger_tableau(donnees_export, paste0(id, "_indicateurs"), "excel")
    output$dl_csv   <- telecharger_tableau(donnees_export, paste0(id, "_indicateurs"), "csv")
    output$dl_pdf_tab <- downloadHandler(
      filename = function() sprintf("%s_indicateurs_%s.pdf", id, format(Sys.Date(), "%Y%m%d")),
      content = function(file) pdf_tableau(file, donnees_export(),
                                           paste("SAD Éducation Togo —", cfg$titre)))
  })
}
